### PR TITLE
#159392862 Create reusable modal component

### PR DESCRIPTION
**Pull request by bahdcoder**
**What does this PR do?**

Create reusable modal & notifications modal components

**Description of Task to be completed?**

 - create and design modal component
 - create and design notifications modal component
 - write tests for modal and notifications modal component
 - add stylings for modal and notifications modal component

**How should this be manually tested?**

- Open the dashboard component. 
- Mount the following code anywhere in the render method: 

```jsx
         <NotificationModal
            showModal
            modalTitle="the test modal title"
            // tslint:disable-next-line
            handleCancel={() => {}}
            // tslint:disable-next-line
            handleAction={() => {}}
            modalContent={{
              heading: 'Are you sure you want to delete ?',
              body: 'This asset has 8 caetgories.',
            }}
          />

```

- Don't forget to import the Notifications component.

**Any background context you want to provide?**

xxx

**What are the relevant pivotal tracker stories?**

[159392862](https://www.pivotaltracker.com/story/show/159392862)

**Screenshots (if appropriate)**

screen shot 2018-07-30 at 9 28 08 am

**Questions:**

xxx

### CODE REVIEW

File: src/components/Modal/Modal.scss [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  top: 0;
+  right: 0;
+  bottom: 0;
+  left: 0;
+  background: map-get($grey, 100);
+  margin: auto;
+  padding: rem(50px);
+  border-radius: rem(4px);
+  box-sizing: border-box;
+  box-shadow: 0 rem(2px) rem(10px) 0 $box-shadow;
```
Comment (Billmike) [inline src/components/Modal/Modal.scss]: This variable is named `$box-shadow` but what we have in the `variables.scss` file is `$boxshadow`

File: src/components/Modal/Modal.scss [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  &--show {
+    display: flex;
+    align-items: center;
+    justify-content: center;
+  }
+}
+
+.modal-dialog {
+  position: fixed;
+  z-index: 99999;
```
Comment (Billmike) [inline src/components/Modal/Modal.scss]: A z-index of 99999? 🤔 I'm curious as to how you arrived at that value for a z-index?

File: src/components/NotificationModal/NotificationModal.test.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
+// react libraries
+import * as React from 'react';
+
+// third-party libraries
+import { mount } from 'enzyme';
+
+
```
Comment (FelixWambiri) [inline src/components/NotificationModal/NotificationModal.test.tsx]: I think you should delete one of the blanks

File: src/components/Modal/Modal.scss (hunk 2 above)
Comment (bahdcoder) [reply src/components/Modal/Modal.scss]: What we want is a z-index value greater than that of all other elements on the DOM.

File: src/components/NotificationModal/NotificationModal.test.tsx (hunk 3 above)
Comment (bahdcoder) [reply src/components/NotificationModal/NotificationModal.test.tsx]: This has been addressed. thanks.

File: src/components/Modal/Modal.scss (hunk 1 above)
Comment (bahdcoder) [reply src/components/Modal/Modal.scss]: This has been addressed. thanks.

File: src/components/Modal/index.tsx [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import './Modal.scss';
+
+// interfaces
+import { ModalProps } from './interfaces';
+
+const Modal = ({ classes, showModal, renderHeader, renderContent, renderFooter }: ModalProps) => ((
+  <React.Fragment>
+    <div className={`${showModal ? 'modal-container--show' : ''} modal-container`}>
+      <div className={`modal-dialog ${classes}`}>
+          <div className="modal-dialog__header">
```
Comment (Dammyy) [inline src/components/Modal/index.tsx]: fix indentation of lines 14 -22

File: src/components/NotificationModal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        <div className="notifications-modal__content">
+          <h3 className="notifications-modal__content__heading">{modalContent.heading}</h3>
+          <p className="notifications-modal__content__body">{modalContent.body}</p>
+        </div>
+      ))}
+      renderFooter={() => ((
+        <div className="notifications-modal__footer">
+          <Button name="Cancel" onClick={handleCancel} />
+          <Button
+            name={action} type="blue regular notifications-modal__footer__action-button"
```
Comment (Billmike) [inline src/components/NotificationModal/index.tsx]: props `type` should go on a new line too

Comment (bahdcoder) [reply src/components/NotificationModal/index.tsx]: @BillmikeI personally do not see any useful reason for this. 🤔  Could you please throw more light? Thanks.

Comment (Billmike) [reply src/components/NotificationModal/index.tsx]: `name` and `onClick` are already on separate lines. It's easier to read if `type` follows suit and takes a new line under `name`.

File: src/components/Modal/Modal.test.tsx [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+  it('adds a show class to modal backdrop when showModal is set to true', () => {
+    expect(wrapper.find('.modal-backdrop--show').length).toBe(1);
+  });
+
+  const hiddenProps = {
+    ...props,
+    showModal: false,
+  };
+  const hiddenWrapper = mount(<Modal { ...hiddenProps } />);
```
Comment (andela-kakpobome) [inline src/components/Modal/Modal.test.tsx]: Please move these to the section where you define other constants

File: src/components/NotificationModal/index.tsx [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+// components
+import Button from 'components/Button';
+import Modal from 'components/Modal';
+
+// interfaces
+import { NotificationModalProps } from './interfaces';
+
+const NotificationModal = ({
+  showModal, modalTitle, handleCancel, handleAction, modalContent, action = 'Delete',
```
Comment (andela-kakpobome) [inline src/components/NotificationModal/index.tsx]: These would read better if they were on separate lines

File: src/components/NotificationModal/index.tsx (hunk 5 above)
Comment (andela-kakpobome) [reply src/components/NotificationModal/index.tsx]: +1 @Billmike

**Comment by andela-kakpobome** [comment]
The showModal prop in your example is not set to any value, and there's no default in NotificationModal, how else would that value be set?

**Comment by bahdcoder** [comment]
@bourgeois247 passing `showModal={true}` is exactly the same as `showModal`. Also, not passing the prop results to a `falsy` value, which yields us the same result as defaulting it to false.

**Comment by andela-kakpobome** [comment]
Makes sense.

**Pull request merged**